trieste/space.py

```diff
@@ -339,3 +339,117 @@ def __mul__(self, other: Box) -> Box:
+        """
+        samples = super().sample(num_samples)
+        samples = tf.round(samples / self._stepsizes) * self._stepsizes
+
+        return samples
+
+    def sample_halton(self, num_samples: int, seed: Optional[int] = None) -> TensorType:
+        """
+        :param num_samples: The number of points to sample from this search space.
+        :return: ``num_samples`` of points, using halton sequence with shape '[num_samples, D]' ,
+            where D is the search space dimension. Rounded by the ``stepsize``.
+        """
+        samples = super().sample_halton(num_samples, seed)
+        samples = tf.round(samples / self._stepsizes) * self._stepsizes
+
+        return samples
+
+    def sample_sobol(self, num_samples: int, skip: Optional[int] = None) -> TensorType:
+        """
+        :param num_samples: The number of points to sample from this search space.
+        :return: ``num_samples`` of points, using sobol sequence with shape '[num_samples, D]' ,
+            where D is the search space dimension. Rounded by the ``stepsize``.
+        """
+        samples = super().sample_sobol(num_samples, skip)
+        samples = tf.round(samples / self._stepsizes) * self._stepsizes
+
+        return samples
+
+    @property
+    def stepsizes(self) -> TensorType:
+        """the step sizes of the ordinal space"""
+        return self._stepsizes
+
+    def __mul__(self, other: OrdinalSearchSpace) -> OrdinalSearchSpace:  # type: ignore[override]
+        r"""
+        Return the Cartesian product of the two :class:`OrdinalSearchSpace`\ es (concatenating
+        their respective lower and upper bounds). For example:
+
+            >>> unit_interval = OrdinalSearchSpace([0.0], [1.0], [0.1])
+            >>> square_at_origin = OrdinalSearchSpace([-2.0, -2.0], [2.0, 2.0], [0.2, 0.3])
+            >>> new_ordinal_space = unit_interval * square_at_origin
+            >>> new_ordinal_space.lower.numpy()
+            array([ 0., -2., -2.])
+            >>> new_ordinal_space.upper.numpy()
+            array([1., 2., 2.])
+            >>> new_ordinal_space.stepsizes.numpy()
+            array([0.1, 0.2, 0.3])
+
+        :param other: A :class:`OrdinalSearchSpace` with bounds of the same type as
+            this :class:`OrdinalSearchSpace`.
+        :return: The Cartesian product of the two :class:`OrdinalSearchSpace`\ es.
+        :raise TypeError: If the bounds of one :class:`OrdinalSearchSpace` have different
+            dtypes to those of the other :class:`OrdinalSearchSpace`.
+        """
+        if self.lower.dtype is not other.lower.dtype:
+            return NotImplemented
+
+        product_lower_bound = tf.concat([self._lower, other.lower], axis=-1)
+        product_upper_bound = tf.concat([self._upper, other.upper], axis=-1)
+        product_stepsizes = tf.concat([self._stepsizes, other.stepsizes], axis=-1)
+
+        return OrdinalSearchSpace(product_lower_bound, product_upper_bound, product_stepsizes)
```

Comment:
hmm it's quite tricky to write this membership function, I am not sure how to check if the points match with the interval since modulo is not working on floating point, any idea?

Reply:
Not sure, you will have to think about it. But it is important to have this functionality. Maybe ask Ivo

Reply:
okay 👌🏼